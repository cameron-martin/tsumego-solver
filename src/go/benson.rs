use super::{BitBoard, GoBoard, GoPlayer};

impl GoBoard {
    fn small_x_enclosed_regions(&self, x: GoPlayer) -> BitBoard {
        let regions = !self.get_bitboard_for_player(x);
        let regions_with_empty_interiors =
            (regions.interior() & self.empty_cells()).flood_fill(regions);

        regions & !regions_with_empty_interiors
    }

    pub fn unconditionally_alive_blocks_for_player(&self, player: GoPlayer) -> BitBoard {
        let mut regions = self.small_x_enclosed_regions(player);
        let blocks = self.get_bitboard_for_player(player);
        let mut remaining_blocks = blocks;

        loop {
            let mut removed_blocks = 0;

            for block in remaining_blocks.iter() {
                // Has the block `block` got at least 2 healthy regions in `regions`?

                let empty_intersections_in_regions = regions & self.empty_cells();

                let unhealthy_regions = (empty_intersections_in_regions
                    & !block.immediate_exterior())
                .flood_fill(regions);

                let healthy_regions = regions & !unhealthy_regions;

                let more_than_one_healthy_region = !healthy_regions.is_empty()
                    && !(healthy_regions
                        & !healthy_regions.first_cell().flood_fill(healthy_regions))
                    .is_empty();

                if !more_than_one_healthy_region {
                    remaining_blocks = remaining_blocks & !block;

                    removed_blocks += 1;
                }
            }

            if removed_blocks == 0 {
                return remaining_blocks;
            }

            let removed_blocks = blocks & !remaining_blocks;
            regions = regions & !(removed_blocks.expand_one() & regions).flood_fill(regions);
        }
    }

    pub fn unconditionally_alive_blocks(&self) -> GoBoard {
        GoBoard {
            white: self.unconditionally_alive_blocks_for_player(GoPlayer::White),
            black: self.unconditionally_alive_blocks_for_player(GoPlayer::Black),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::super::GoGame;
    use super::*;

    #[test]
    fn small_black_enclosed_regions() {
        let board = GoGame::from_sgf(include_str!(
            "../test_sgfs/small_black_enclosed_regions.sgf"
        ))
        .get_board();

        let answer = GoGame::from_sgf(include_str!(
            "../test_sgfs/small_black_enclosed_regions_answer.sgf"
        ))
        .get_board()
        .get_bitboard_for_player(GoPlayer::White);

        assert_eq!(board.small_x_enclosed_regions(GoPlayer::Black), answer);
    }

    #[test]
    fn all_alive1() {
        let game = GoGame::from_sgf(include_str!("../test_sgfs/life_and_death/all_alive1.sgf"));

        assert_eq!(
            game.get_board()
                .unconditionally_alive_blocks_for_player(GoPlayer::White),
            game.get_board().get_bitboard_for_player(GoPlayer::White)
        );
    }

    #[test]
    fn all_dead1() {
        let game = GoGame::from_sgf(include_str!("../test_sgfs/life_and_death/all_dead1.sgf"));

        assert_eq!(
            game.get_board().unconditionally_alive_blocks(),
            GoBoard::empty(),
        );
    }

    #[test]
    fn mixture() {
        let game = GoGame::from_sgf(include_str!("../test_sgfs/life_and_death/mixture.sgf"));
        let answer = GoGame::from_sgf(include_str!(
            "../test_sgfs/life_and_death/mixture_answer.sgf"
        ));

        assert_eq!(
            game.get_board().unconditionally_alive_blocks(),
            *answer.get_board(),
        );
    }
}
