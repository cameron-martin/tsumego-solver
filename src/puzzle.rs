use crate::go::{GoGame, GoPlayer};
use petgraph::graph::NodeIndex;
use petgraph::{Direction, Graph};
use std::fmt;
use std::fmt::{Debug, Formatter};
use std::ops::Add;

#[derive(PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
pub enum ProofNumber {
    Finite(u32),
    Infinity,
}

impl Debug for ProofNumber {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        match self {
            ProofNumber::Finite(n) => Debug::fmt(n, f),
            ProofNumber::Infinity => f.write_str("âˆž"),
        }
    }
}

impl Add for ProofNumber {
    type Output = Self;

    fn add(self, rhs: Self) -> Self {
        match (self, rhs) {
            (ProofNumber::Finite(n), ProofNumber::Finite(m)) => ProofNumber::Finite(n + m),
            _ => ProofNumber::Infinity,
        }
    }
}

#[derive(PartialEq, Clone, Copy)]
pub enum NodeValue {
    True,
    False,
    Unknown,
}

impl Debug for NodeValue {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_str(match self {
            NodeValue::False => "False",
            NodeValue::True => "True",
            NodeValue::Unknown => "Unknown",
        })
    }
}

pub enum NodeType {
    And,
    Or,
}

impl Debug for NodeType {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_str(match self {
            NodeType::And => "And",
            NodeType::Or => "Or",
        })
    }
}

pub struct AndOrNode {
    of_type: NodeType,
    value: NodeValue,
    proof_number: ProofNumber,
    disproof_number: ProofNumber,
    game: GoGame,
}

impl Debug for AndOrNode {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        f.write_fmt(format_args!(
            "{:?} {:?} {:?}/{:?}",
            self.of_type, self.value, self.proof_number, self.disproof_number
        ))
    }
}

impl AndOrNode {
    pub fn create_unknown_leaf(of_type: NodeType, game: GoGame) -> AndOrNode {
        AndOrNode {
            of_type,
            value: NodeValue::Unknown,
            proof_number: ProofNumber::Finite(1),
            disproof_number: ProofNumber::Finite(1),
            game,
        }
    }

    pub fn create_true_leaf(of_type: NodeType, game: GoGame) -> AndOrNode {
        AndOrNode {
            of_type,
            value: NodeValue::True,
            proof_number: ProofNumber::Finite(0),
            disproof_number: ProofNumber::Infinity,
            game,
        }
    }

    pub fn create_false_leaf(of_type: NodeType, game: GoGame) -> AndOrNode {
        AndOrNode {
            of_type,
            value: NodeValue::False,
            proof_number: ProofNumber::Infinity,
            disproof_number: ProofNumber::Finite(0),
            game,
        }
    }

    pub fn is_proved(&self) -> bool {
        self.proof_number == ProofNumber::Finite(0)
    }

    pub fn is_disproved(&self) -> bool {
        self.disproof_number == ProofNumber::Finite(0)
    }
}

pub struct Puzzle {
    player: GoPlayer,
    attacker: GoPlayer,
    tree: Graph<AndOrNode, ()>,
    root_id: NodeIndex,
}

impl Puzzle {
    pub fn new(game: GoGame, attacker: GoPlayer) -> Puzzle {
        // debug_assert_eq!(game.plys(), 0);

        let player = game.current_player;

        let mut tree = Graph::new();

        let root_id = tree.add_node(AndOrNode::create_unknown_leaf(NodeType::Or, game));

        Puzzle {
            player,
            attacker,
            tree,
            root_id,
        }
    }

    fn defender(&self) -> GoPlayer {
        self.attacker.flip()
    }

    fn develop_node(&mut self, node_id: NodeIndex) {
        debug_assert!(self.tree.neighbors(node_id).next().is_none());
        debug_assert_eq!(self.tree[node_id].value, NodeValue::Unknown);

        let mut any_false = false;
        let mut any_true = false;
        let mut any_unknown = false;

        let node = &self.tree[node_id];

        let mut moves = node.game.generate_moves();

        if !node.game.last_move_pass {
            moves.push(node.game.pass());
        }

        debug_assert_ne!(
            moves.len(),
            0,
            "{:?} {:?} to play\n{:?}",
            node,
            node.game.current_player,
            node.game.get_board()
        );

        for child in moves {
            let node = &self.tree[node_id];

            let node_type = if child.current_player == self.player {
                NodeType::Or
            } else {
                NodeType::And
            };

            let new_node = if !node
                .game
                .get_board()
                .unconditionally_alive_blocks_for_player(self.defender())
                .is_empty()
            {
                if self.defender() == self.player {
                    AndOrNode::create_true_leaf(node_type, child)
                } else {
                    AndOrNode::create_false_leaf(node_type, child)
                }
            } else if false {
                // TODO: Work out how we lose
                panic!()
            } else {
                AndOrNode::create_unknown_leaf(node_type, child)
            };

            match new_node.value {
                NodeValue::True => any_true = true,
                NodeValue::False => any_false = true,
                NodeValue::Unknown => any_unknown = true,
            }

            let new_node_id = self.tree.add_node(new_node);

            self.tree.add_edge(node_id, new_node_id, ());
        }

        let node = &mut self.tree[node_id];

        node.value = match node.of_type {
            NodeType::And => {
                if any_false {
                    NodeValue::False
                } else if any_unknown {
                    NodeValue::Unknown
                } else {
                    NodeValue::True
                }
            }
            NodeType::Or => {
                if any_true {
                    NodeValue::True
                } else if any_unknown {
                    NodeValue::Unknown
                } else {
                    NodeValue::False
                }
            }
        }
    }

    fn select_most_proving_node(&self, start_node_id: NodeIndex) -> NodeIndex {
        let mut current_node_id = start_node_id;

        loop {
            let node = &self.tree[current_node_id];

            let mut neighbours = self.tree.neighbors(current_node_id);

            if neighbours.clone().next().is_none() {
                debug_assert_eq!(
                    node.value,
                    NodeValue::Unknown,
                    "{:?} {:?} to play\n{:?}",
                    node,
                    node.game.current_player,
                    node.game.get_board()
                );
                break;
            }

            current_node_id = match node.of_type {
                NodeType::Or => {
                    debug_assert_ne!(
                        node.proof_number,
                        ProofNumber::Finite(0),
                        "{:?} {:?} to play\n{:?}",
                        node,
                        node.game.current_player,
                        node.game.get_board()
                    );

                    neighbours
                        .find(|&child_id| {
                            let child = &self.tree[child_id];

                            child.proof_number == node.proof_number
                        })
                        .unwrap()
                }
                NodeType::And => {
                    debug_assert_ne!(
                        node.disproof_number,
                        ProofNumber::Finite(0),
                        "{:?} {:?} to play\n{:?}",
                        node,
                        node.game.current_player,
                        node.game.get_board()
                    );

                    neighbours
                        .find(|&child_id| {
                            let child = &self.tree[child_id];

                            child.disproof_number == node.disproof_number
                        })
                        .unwrap()
                }
            }
        }

        current_node_id
    }

    pub fn solve(&mut self) {
        while !self.is_solved() {
            self.solve_iteration();
        }
    }

    fn is_solved(&self) -> bool {
        self.root_node().is_proved() || self.root_node().is_disproved()
    }

    fn solve_iteration(&mut self) {
        let most_proving_node_id = self.select_most_proving_node(self.root_id);

        self.develop_node(most_proving_node_id);
        self.update_ancestors(most_proving_node_id);
    }

    fn update_ancestors(&mut self, node_id: NodeIndex) {
        let mut current_node_id = node_id;

        loop {
            self.set_proof_and_disproof_numbers(current_node_id);

            if let Some(parent_node_id) = self
                .tree
                .neighbors_directed(current_node_id, Direction::Incoming)
                .next()
            {
                current_node_id = parent_node_id;
            } else {
                break;
            }
        }
    }

    fn set_proof_and_disproof_numbers(&mut self, node_id: NodeIndex) {
        let children = self
            .tree
            .neighbors(node_id)
            .map(|child_id| &self.tree[child_id]);

        let mut proof_number_sum = ProofNumber::Finite(0);
        let mut proof_number_min = ProofNumber::Infinity;
        let mut disproof_number_sum = ProofNumber::Finite(0);
        let mut disproof_number_min = ProofNumber::Infinity;

        for child in children {
            proof_number_sum = proof_number_sum + child.proof_number;
            disproof_number_sum = disproof_number_sum + child.disproof_number;

            if child.proof_number < proof_number_min {
                proof_number_min = child.proof_number;
            }

            if child.disproof_number < disproof_number_min {
                disproof_number_min = child.disproof_number;
            }
        }

        let node = &mut self.tree[node_id];
        match node.of_type {
            NodeType::And => {
                node.proof_number = proof_number_sum;
                node.disproof_number = disproof_number_min;
            }
            NodeType::Or => {
                node.proof_number = proof_number_min;
                node.disproof_number = disproof_number_sum;
            }
        }
    }

    fn root_node(&self) -> &AndOrNode {
        &self.tree[self.root_id]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::go::GoGame;

    #[test]
    fn true_simple1() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple1.sgf"));

        let mut puzzle = Puzzle::new(tsumego, GoPlayer::White);

        puzzle.solve();

        assert!(puzzle.root_node().is_proved());
    }

    #[test]
    fn true_simple2() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple2.sgf"));

        let mut puzzle = Puzzle::new(tsumego, GoPlayer::Black);

        puzzle.solve();

        assert!(puzzle.root_node().is_proved());
    }

    #[test]
    fn proof_number_ordering() {
        assert!(ProofNumber::Infinity > ProofNumber::Finite(1));
    }
}
