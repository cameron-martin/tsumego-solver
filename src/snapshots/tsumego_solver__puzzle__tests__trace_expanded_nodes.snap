---
source: src/puzzle.rs
expression: output
---
. . . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b . . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. . b w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. . . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

w b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b b . b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

w b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

w . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b . . b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b b w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b . b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

w . . w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b b w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b w b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b . b _ _ _ _ _ _ _ _ _ _ _
w . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b . b _ _ _ _ _ _ _ _ _ _ _
. w . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b . b _ _ _ _ _ _ _ _ _ _ _
. . w . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b . b _ _ _ _ _ _ _ _ _ _ _
. . . w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b b b . b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b . . b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

w . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. w w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b . w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. b w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

b . . . b _ _ _ _ _ _ _ _ _ _ _
. . . . b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

. w w w b _ _ _ _ _ _ _ _ _ _ _
w w w w b _ _ _ _ _ _ _ _ _ _ _
b b b b b _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


